linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -2624,6 +2624,86 @@ public static boolean isMergeJoinSupported(JoinType joinType) {
     };
   }
 
+
+  /**
+   * A sort implementation optimized for a sort with a fetch size (LIMIT).
+   * @param offset how many rows are skipped from the sorted output.
+   *               Must be greater than or equal to 0.
+   * @param fetch how many rows are retrieved. Must be greater than 0.
+   */
+  public static <TSource, TKey> Enumerable<TSource> orderBy(
+      Enumerable<TSource> source,
+      Function1<TSource, TKey> keySelector,
+      Comparator<TKey> comparator,
+      int offset, int fetch) {
+    return new AbstractEnumerable<TSource>() {
+      @Override public Enumerator<TSource> enumerator() {
+        TreeMap<TKey, List<TSource>> map = new TreeMap<>(comparator);
+        long size = 0;
+        long needed = fetch + offset;
+
+        try (Enumerator<TSource> os = source.enumerator()) {
+          while (os.moveNext()) {
+            TSource o = os.current();
+            TKey key = keySelector.apply(o);
+            if (needed >= 0 && size >= needed) {
+              if (comparator.compare(key, map.lastKey()) >= 0) {
+                continue;
+              }
+              // remove last entry from tree map
+              List<TSource> l = map.get(map.lastKey());
+              if (l.size() == 1) {
+                map.remove(map.lastKey());
+              } else {
+                l.remove(l.size() - 1);
+              }
+              size--;
+            }
+            map.compute(key, (k, l) -> {
+              if (l == null) {
+                return Collections.singletonList(o);
```

Comment:
minor: maybe adding here the same comment that we have in `toLookup_` to explain the choice of List implementation?
`// for first entry, use a singleton list to save space`
`// when we go from 1 to 2 elements, switch to array list`

Reply:
+1. I was also looking for some comments to explain the algorithm briefly.

Reply:
@amaliujia, do you mean that I should add some comments that explain the implementation of the new orderBy method.

Reply:
Yes. 

Or you can link the JIRA but update the JIRA with the final detailed algorithm you have chosen. 

I found the JIRA has many discussions so was confused on which algorithm you finally chose, until I read the implementation here :). Something help people easy to understand the implementation in the future will be very helpful.

Reply:
I've added an explanation at the beginning and more comments in the actual code to make the algorithm easier to understand.